// @mintlify/mcp/src/mcp-server/src/utils.ts
import { OpenAPI } from '@mintlify/openapi-types';
import axios, { AxiosResponse } from 'axios';

export type NestedRecord = string | { [key: string]: NestedRecord };
export type SimpleRecord = Record<string, NestedRecord>;

export function initializeObject(obj: SimpleRecord, path: string[]): SimpleRecord {
  let current = obj;
  for (const key of path) {
    if (!current[key] || typeof current[key] !== 'object') {
      current[key] = {};
    }
    current = current[key];
  }
  return current;
}

export function getFileId(spec: OpenAPI.Document, index: number): string | number {
  return spec.info?.title && spec.info.version
    ? `${spec.info.title} - ${spec.info.version}`
    : index;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function throwOnAxiosError(response: AxiosResponse<any, any>, errMsg: string) {
  if (response.status !== 200) {
    if (response.headers['content-type']?.includes('application/json') && response.data?.error) {
      throw new Error(`${errMsg}: ${response.data.error}`);
    } else {
      throw new Error(`${errMsg}: ${response.status} ${response.statusText || ''}`);
    }
  }

  if (!response.data) {
    throw new Error(`${errMsg}: ${response.status} ${response.statusText}`);
  }
}

export function formatErr(err: unknown) {
  if (axios.isAxiosError(err)) {
    if (err.message) {
      return err.message;
    } else if (err.response) {
      return err.response.data?.error ?? `${err.response.status} ${err.response.statusText}`;
    } else if (err.request) {
      return 'No response received from server';
    } else {
      err = 'An unknown error occurred';
    }
  } else if (err instanceof Error) {
    return err.message;
  } else {
    return JSON.stringify(err, undefined, 2);
  }
}
