// @mintlify/mcp/src/utils/cli-config.ts
import { colors } from '@mintlify/scraping';
import { DataSchemaArray, Endpoint, SecurityParameterSchema } from '@mintlify/validation';
import fs from 'fs';
import path from 'path';
import prompts from 'prompts';

import { fetchSearchConfigurationAndTools } from '../mcp-server/src/search.js';
import { convertStrToTitle as conv } from '../mcp-server/src/tools/helpers.js';
import { InitializationConfiguration, ToolWithEndpointWithUuid } from '../mcp-server/src/types.js';
import { initializeObject, SimpleRecord } from '../mcp-server/src/utils.js';
import { McpMetadata } from '../remote.js';
import { generatePackageJson, onCancel } from '../utils.js';

export async function buildMcp(mcpDir: string, subdomain: string, filename: string): Promise<void> {
  let config: InitializationConfiguration;
  try {
    config = await fetchSearchConfigurationAndTools(subdomain);
  } catch (err) {
    throw err;
  }

  if (fs.existsSync(mcpDir)) {
    fs.rmSync(mcpDir, { recursive: true, force: true });
  }
  fs.mkdirSync(mcpDir, { recursive: true });
  const packageJson = generatePackageJson(subdomain);
  fs.writeFileSync(path.join(mcpDir, 'package.json'), JSON.stringify(packageJson, undefined, 2));
  const cpPath = path.join(path.dirname(filename), 'mcp-server');
  fs.cpSync(cpPath, mcpDir, { recursive: true });
  const configFile = path.join(mcpDir, 'src', 'config.readonly.js');
  const configContent = fs.readFileSync(configFile, 'utf8');
  const updatedConfigContent = configContent.replace('mintlify', subdomain);
  fs.writeFileSync(configFile, updatedConfigContent);

  if (config.tools && config.tools.length) {
    await initializeToolsFromConfig(config.tools, mcpDir);
  }
}

export async function getEnvFromMetadata(
  mcpMetadata: McpMetadata,
  subdomain: string
): Promise<Record<string, string> | undefined> {
  let envVars: Record<string, string> | undefined = undefined;
  for (const key of Object.keys(mcpMetadata.mcpConfiguration[subdomain]?.env ?? {})) {
    const envResponse = await prompts(
      {
        type: 'invisible',
        name: key,
        message: ` ${colors.dim}What is the ${colors.default}${colors.green}${key}${colors.default}${colors.dim} for ${colors.default}"${subdomain}"`,
      },
      {
        onCancel,
      }
    );
    if (!envVars) {
      envVars = {};
    }
    envVars[key] = envResponse[key];
  }
  return envVars;
}

async function initializeToolsFromConfig(
  tools: ToolWithEndpointWithUuid[],
  mcpDir: string
): Promise<void> {
  console.log(`🛠  ${colors.bold}Installing Tools:${colors.default}`);
  console.log();
  for (const tool of tools) {
    console.log(`   ${tool.tool.name}${colors.default}`);
    console.log(`   ${colors.dim}${tool.tool.description}${colors.default}`);
    console.log();
  }

  fs.writeFileSync(path.join(mcpDir, 'src', 'tools.json'), JSON.stringify(tools, null, 2));

  const envVars: SimpleRecord = {};

  for (const tool of tools) {
    if (tool.uuid) {
      await collectEnvVars(tool.endpoint, tool.uuid, tool.fileName, envVars);
    }
  }

  if (Object.keys(envVars).length) {
    writeEnvFile(mcpDir, envVars);
  }
}

async function collectEnvVars(
  endpoint: Endpoint<DataSchemaArray>,
  currFileId: string | number,
  currFileName: string,
  envVars: SimpleRecord
): Promise<void> {
  // If the env vars for this file have already been collected, skip
  if (envVars[currFileId] !== undefined) {
    return;
  }

  envVars[currFileId] = {};
  const targetEnv = envVars[currFileId] as SimpleRecord;

  const security = endpoint.request.security;
  for (const option of security) {
    for (const [location, group] of Object.entries(option.parameters)) {
      for (const [name, param] of Object.entries(group)) {
        const promptKey = `${endpoint.method}-${endpoint.path}-${name}`;
        await handleSecurityParameter(param, location, name, currFileName, promptKey, targetEnv);
      }
    }
  }
}

async function handleSecurityParameter(
  param: SecurityParameterSchema,
  location: string,
  name: string,
  currFileName: string,
  promptKey: string,
  targetEnv: SimpleRecord
): Promise<void> {
  switch (param.type) {
    case 'apiKey':
      await handleApiKeyParameter(location, name, currFileName, promptKey, targetEnv);
      break;
    case 'http':
      await handleHttpParameter(param, location, name, currFileName, promptKey, targetEnv);
      break;
    case 'oauth2':
    default:
      break;
  }
}

async function handleApiKeyParameter(
  location: string,
  name: string,
  currFileName: string,
  promptKey: string,
  targetEnv: SimpleRecord
): Promise<void> {
  const apiKeyTarget = initializeObject(targetEnv, [location, name]);
  if (apiKeyTarget['API_KEY'] !== undefined) return;

  console.log();
  const apiKeyResponse = await prompts(
    {
      type: 'invisible',
      name: promptKey,
      message: ` ${colors.bold}What is the ${colors.green}API Key${colors.default}${colors.bold}${
        colors.dim
      }${typeof currFileName === 'string' ? ` for ${colors.default}"${currFileName}"` : ''}${
        colors.default
      }?`,
    },
    {
      onCancel,
    }
  );

  const apiKey = apiKeyResponse[promptKey];
  apiKeyTarget['API_KEY'] = apiKey;
}

async function handleHttpParameter(
  param: {
    type: 'http';
    scheme: 'bearer' | 'basic';
    description?: string;
  },
  location: string,
  name: string,
  currFileName: string,
  promptKey: string,
  targetEnv: SimpleRecord
): Promise<void> {
  const httpTarget = initializeObject(targetEnv, [location, name, 'HTTP']);
  if (httpTarget[param.scheme] !== undefined) return;

  console.log();
  const httpResponse = await prompts(
    {
      type: 'invisible',
      name: promptKey,
      message: ` ${colors.bold}What is the ${colors.green}${conv(param.scheme)} Token${
        colors.default
      }${colors.bold}${typeof currFileName === 'string' ? ` for ${currFileName}` : ''}${
        colors.default
      }?`,
    },
    {
      onCancel,
    }
  );

  const token = httpResponse[promptKey];
  httpTarget[param.scheme] = token?.startsWith(`${param.scheme} `)
    ? token
    : `${param.scheme} ${token}`;
}

function writeEnvFile(mcpDir: string, envVars: SimpleRecord): void {
  const newEnvContent = JSON.stringify(envVars, undefined, 2);
  const envPath = path.join(mcpDir, '.env.json');
  fs.writeFileSync(envPath, newEnvContent);
}
