// @mintlify/mcp/src/utils/markdown.ts
import { colors, Color } from '@mintlify/scraping';
import { remark } from 'remark';
import type { Node } from 'unist';

import { getTerminalWidth, padString, wrapText } from '../utils.js';

export function colorize(text: string, ...colorCodes: Color[]): string {
  const start = colorCodes.map((code) => colors[code]).join('');
  return `${start}${text}${colors.default}`;
}

type MarkdownNode = Node & {
  type: string;
  value?: string;
  children?: MarkdownNode[];
  ordered?: boolean;
  depth?: number;
  url?: string;
  alt?: string;
  lang?: string;
  checked?: boolean;
};

const processor = remark();

export function renderMarkdown(markdown: string): string {
  const ast = processor.parse(markdown);

  const terminalWidth = getTerminalWidth();
  const boxWidth = terminalWidth - 4;
  const contentWidth = boxWidth - 2;

  const rendered = renderNode(ast as MarkdownNode, 0);
  const paragraphs = rendered.split('\n\n');

  const topBorder = '┌' + '─'.repeat(boxWidth) + '┐';
  const bottomBorder = '└' + '─'.repeat(boxWidth) + '┘';
  let result = topBorder + '\n';

  for (let i = 0; i < paragraphs.length; i++) {
    const paragraph = paragraphs[i] ?? '';
    if (paragraph === '') continue;

    if (paragraph.includes('•') || /^\d+\./.test(paragraph)) {
      const listLines = paragraph.split('\n');
      for (const line of listLines) {
        const wrappedLines = wrapText(line, contentWidth);
        for (const wrappedLine of wrappedLines) {
          result += `│ ${padString(wrappedLine ?? '', contentWidth)} │\n`;
        }
      }
    } else {
      const wrappedLines = wrapText(paragraph, contentWidth);
      for (const line of wrappedLines) {
        result += `│ ${padString(line ?? '', contentWidth)} │\n`;
      }
    }

    if (i < paragraphs.length - 1) {
      result += `│ ${' '.repeat(contentWidth)} │\n`;
    }
  }

  result += bottomBorder;
  return result;
}

function renderNode(node: MarkdownNode, depth: number = 0): string {
  switch (node.type) {
    case 'root':
      return node.children?.map((child) => renderNode(child, depth)).join('') || '';

    case 'heading':
      return (
        colorize(
          node.children?.map((child) => renderNode(child, depth)).join('') || '',
          'bold',
          'green'
        ) + '\n\n'
      );

    case 'paragraph':
      return (
        colorize(
          node.children?.map((child) => renderNode(child, depth)).join('') || '',
          'default'
        ) + '\n\n'
      );

    case 'strong':
      return colorize(
        node.children?.map((child) => renderNode(child, depth)).join('') || '',
        'bold'
      );

    case 'emphasis':
      return colorize(
        node.children?.map((child) => renderNode(child, depth)).join('') || '',
        'italic'
      );

    case 'link':
      return node.children?.map((child) => renderNode(child, depth)).join('') || '';

    case 'text':
      return node.value || '';

    case 'list':
      const items =
        node.children
          ?.map((item, index) => {
            const bullet = node.ordered ? `${index + 1}.` : '•';
            const indent = ' '.repeat(depth * 2);
            const itemContent = renderNode(item, depth + 1);

            return `${indent}${colorize(bullet, 'default')} ${itemContent}`;
          })
          .join('') || '';
      return items;

    case 'listItem':
      return node.children?.map((child) => renderNode(child, depth)).join('') || '';

    case 'code':
      return colorize(node.value || '', 'default') + '\n\n';

    case 'blockquote':
      const quoteContent =
        node.children?.map((child) => renderNode(child, depth + 1)).join('') || '';
      const indentedQuote = quoteContent
        .split('\n')
        .map((line) => `> ${line}`)
        .join('\n');
      return colorize(indentedQuote, 'default') + '\n';

    case 'thematicBreak':
      return colorize('---', 'default') + '\n\n';

    case 'image':
      return colorize(`[Image: ${node.alt || 'image'}]`, 'default');

    case 'inlineCode':
      return colorize(node.value || '', 'default');

    default:
      if (node.children && node.children.length > 0) {
        return node.children.map((child) => renderNode(child, depth)).join('');
      }
      if (node.value) {
        return colorize(node.value, 'default');
      }
      return '';
  }
}
