// @mintlify/mcp/bin/utils/cli-config.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { colors } from '@mintlify/scraping';
import fs from 'fs';
import path from 'path';
import prompts from 'prompts';
import { fetchSearchConfigurationAndTools } from '../mcp-server/src/search.js';
import { convertStrToTitle as conv } from '../mcp-server/src/tools/helpers.js';
import { initializeObject } from '../mcp-server/src/utils.js';
import { generatePackageJson, onCancel } from '../utils.js';
export function buildMcp(mcpDir, subdomain, filename) {
    return __awaiter(this, void 0, void 0, function* () {
        let config;
        try {
            config = yield fetchSearchConfigurationAndTools(subdomain);
        }
        catch (err) {
            throw err;
        }
        if (fs.existsSync(mcpDir)) {
            fs.rmSync(mcpDir, { recursive: true, force: true });
        }
        fs.mkdirSync(mcpDir, { recursive: true });
        const packageJson = generatePackageJson(subdomain);
        fs.writeFileSync(path.join(mcpDir, 'package.json'), JSON.stringify(packageJson, undefined, 2));
        const cpPath = path.join(path.dirname(filename), 'mcp-server');
        fs.cpSync(cpPath, mcpDir, { recursive: true });
        const configFile = path.join(mcpDir, 'src', 'config.readonly.js');
        const configContent = fs.readFileSync(configFile, 'utf8');
        const updatedConfigContent = configContent.replace('mintlify', subdomain);
        fs.writeFileSync(configFile, updatedConfigContent);
        if (config.tools && config.tools.length) {
            yield initializeToolsFromConfig(config.tools, mcpDir);
        }
    });
}
export function getEnvFromMetadata(mcpMetadata, subdomain) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        let envVars = undefined;
        for (const key of Object.keys((_b = (_a = mcpMetadata.mcpConfiguration[subdomain]) === null || _a === void 0 ? void 0 : _a.env) !== null && _b !== void 0 ? _b : {})) {
            const envResponse = yield prompts({
                type: 'invisible',
                name: key,
                message: ` ${colors.dim}What is the ${colors.default}${colors.green}${key}${colors.default}${colors.dim} for ${colors.default}"${subdomain}"`,
            }, {
                onCancel,
            });
            if (!envVars) {
                envVars = {};
            }
            envVars[key] = envResponse[key];
        }
        return envVars;
    });
}
function initializeToolsFromConfig(tools, mcpDir) {
    return __awaiter(this, void 0, void 0, function* () {
        console.log(`🛠  ${colors.bold}Installing Tools:${colors.default}`);
        console.log();
        for (const tool of tools) {
            console.log(`   ${tool.tool.name}${colors.default}`);
            console.log(`   ${colors.dim}${tool.tool.description}${colors.default}`);
            console.log();
        }
        fs.writeFileSync(path.join(mcpDir, 'src', 'tools.json'), JSON.stringify(tools, null, 2));
        const envVars = {};
        for (const tool of tools) {
            if (tool.uuid) {
                yield collectEnvVars(tool.endpoint, tool.uuid, tool.fileName, envVars);
            }
        }
        if (Object.keys(envVars).length) {
            writeEnvFile(mcpDir, envVars);
        }
    });
}
function collectEnvVars(endpoint, currFileId, currFileName, envVars) {
    return __awaiter(this, void 0, void 0, function* () {
        // If the env vars for this file have already been collected, skip
        if (envVars[currFileId] !== undefined) {
            return;
        }
        envVars[currFileId] = {};
        const targetEnv = envVars[currFileId];
        const security = endpoint.request.security;
        for (const option of security) {
            for (const [location, group] of Object.entries(option.parameters)) {
                for (const [name, param] of Object.entries(group)) {
                    const promptKey = `${endpoint.method}-${endpoint.path}-${name}`;
                    yield handleSecurityParameter(param, location, name, currFileName, promptKey, targetEnv);
                }
            }
        }
    });
}
function handleSecurityParameter(param, location, name, currFileName, promptKey, targetEnv) {
    return __awaiter(this, void 0, void 0, function* () {
        switch (param.type) {
            case 'apiKey':
                yield handleApiKeyParameter(location, name, currFileName, promptKey, targetEnv);
                break;
            case 'http':
                yield handleHttpParameter(param, location, name, currFileName, promptKey, targetEnv);
                break;
            case 'oauth2':
            default:
                break;
        }
    });
}
function handleApiKeyParameter(location, name, currFileName, promptKey, targetEnv) {
    return __awaiter(this, void 0, void 0, function* () {
        const apiKeyTarget = initializeObject(targetEnv, [location, name]);
        if (apiKeyTarget['API_KEY'] !== undefined)
            return;
        console.log();
        const apiKeyResponse = yield prompts({
            type: 'invisible',
            name: promptKey,
            message: ` ${colors.bold}What is the ${colors.green}API Key${colors.default}${colors.bold}${colors.dim}${typeof currFileName === 'string' ? ` for ${colors.default}"${currFileName}"` : ''}${colors.default}?`,
        }, {
            onCancel,
        });
        const apiKey = apiKeyResponse[promptKey];
        apiKeyTarget['API_KEY'] = apiKey;
    });
}
function handleHttpParameter(param, location, name, currFileName, promptKey, targetEnv) {
    return __awaiter(this, void 0, void 0, function* () {
        const httpTarget = initializeObject(targetEnv, [location, name, 'HTTP']);
        if (httpTarget[param.scheme] !== undefined)
            return;
        console.log();
        const httpResponse = yield prompts({
            type: 'invisible',
            name: promptKey,
            message: ` ${colors.bold}What is the ${colors.green}${conv(param.scheme)} Token${colors.default}${colors.bold}${typeof currFileName === 'string' ? ` for ${currFileName}` : ''}${colors.default}?`,
        }, {
            onCancel,
        });
        const token = httpResponse[promptKey];
        httpTarget[param.scheme] = (token === null || token === void 0 ? void 0 : token.startsWith(`${param.scheme} `))
            ? token
            : `${param.scheme} ${token}`;
    });
}
function writeEnvFile(mcpDir, envVars) {
    const newEnvContent = JSON.stringify(envVars, undefined, 2);
    const envPath = path.join(mcpDir, '.env.json');
    fs.writeFileSync(envPath, newEnvContent);
}
