// @mintlify/mcp/bin/utils/mcp-configs.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { colors, log } from '@mintlify/scraping';
import fs from 'fs';
import os from 'os';
import path from 'path';
import prompts from 'prompts';
import { LOGGING_CONFIG, onCancel } from '../utils.js';
const availableConfigMap = {
    cursor: 'Cursor',
    claude: 'Claude Desktop',
    windsurf: 'Windsurf',
};
const ALL_TITLE = 'All';
export function getConfigObj(execPath, env, omitEnvVals) {
    const isBuilt = execPath.endsWith('js');
    return {
        command: isBuilt ? 'node' : 'npx',
        args: isBuilt ? [execPath] : ['tsx', execPath],
        env: omitEnvVals
            ? env
                ? Object.fromEntries(Object.entries(env).map(([key]) => [key, '******']))
                : undefined
            : env,
    };
}
export function modifyMcpConfig(configDir, filename, subdomain, execPath, env) {
    return __awaiter(this, void 0, void 0, function* () {
        const configPath = path.join(configDir, filename);
        try {
            if (!fs.existsSync(configDir)) {
                return false;
            }
            const rawConfig = fs.existsSync(configPath) ? fs.readFileSync(configPath, 'utf8') : '{}';
            const config = JSON.parse(rawConfig);
            if (config.mcpServers === undefined) {
                config.mcpServers = {};
            }
            config.mcpServers[subdomain] = getConfigObj(execPath, env);
            fs.writeFileSync(configPath, JSON.stringify(config, undefined, 2), 'utf8');
            return true;
        }
        catch (err) {
            log(`Error parsing config at ${colors.red}${configPath}${colors.default}: ${err instanceof Error ? err.message : JSON.stringify(err, undefined, 2)}`, 'error', LOGGING_CONFIG);
            return false;
        }
    });
}
function updateConfig(config, subdomain, execPath, env) {
    return __awaiter(this, void 0, void 0, function* () {
        yield modifyMcpConfig(config.dir, config.filename, subdomain, execPath, env);
        return 1;
    });
}
export function modifyAllMcpConfigs(subdomain, execPath, configsToModify, env) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const configPaths = [
                {
                    dir: path.join(os.homedir(), '.cursor'),
                    filename: 'mcp.json',
                    name: 'Cursor',
                },
                {
                    dir: path.join(os.homedir(), '.codeium', 'windsurf'),
                    filename: 'mcp_config.json',
                    name: 'Windsurf',
                },
                {
                    dir: getClaudeConfigPath(),
                    filename: 'claude_desktop_config.json',
                    name: 'Claude Desktop',
                },
            ];
            const existingConfigs = configPaths.filter((config) => fs.existsSync(config.dir));
            if (configsToModify) {
                configsToModify = configsToModify.filter((config) => {
                    const configName = availableConfigMap[config];
                    if (!existingConfigs.find((existingConfig) => existingConfig.name === configName)) {
                        log(`Failed to modify config for ${configName}. ${configName} has not been installed`, 'error', LOGGING_CONFIG);
                        return false;
                    }
                    return true;
                });
            }
            if (existingConfigs.length === 0 || (configsToModify && configsToModify.length === 0)) {
                log('No supported MCP configs found on this system', 'info');
                return;
            }
            const choices = existingConfigs.map((config) => ({
                title: `${config.name}`,
                value: [config],
                selected: true,
            }));
            let response;
            if (configsToModify) {
                response = {
                    selectedConfigs: configPaths.filter((config) => configsToModify === null || configsToModify === void 0 ? void 0 : configsToModify.find((configToModify) => availableConfigMap[configToModify] === config.name)),
                };
            }
            else if (choices.length > 1) {
                console.log();
                response = yield prompts({
                    type: 'select',
                    name: 'selectedConfigs',
                    message: ' Select MCP client:',
                    choices: [
                        ...choices,
                        {
                            title: ALL_TITLE,
                            value: [...existingConfigs],
                            selected: true,
                        },
                    ],
                    hint: 'Use arrow keys to select. Return to submit.',
                    instructions: false,
                }, {
                    onCancel,
                });
            }
            else if (choices[0]) {
                console.log();
                response = yield prompts({
                    type: 'confirm',
                    name: 'selectedConfigs',
                    message: ` Add MCP server to ${choices[0].title} config?`,
                    initial: true,
                }, {
                    onCancel,
                });
            }
            if (!response ||
                !response.selectedConfigs ||
                (Array.isArray(response.selectedConfigs) && response.selectedConfigs.length === 0) ||
                !choices[0]) {
                log('No clients selected', 'warn', LOGGING_CONFIG);
                return;
            }
            if (typeof response.selectedConfigs === 'boolean') {
                response.selectedConfigs = choices[0].value;
            }
            let numUpdates = 0;
            const selectedConfigs = response.selectedConfigs;
            for (const config of selectedConfigs) {
                numUpdates += yield updateConfig(config, subdomain, execPath, env);
            }
            const namesOfSelectedConfigs = selectedConfigs.map((config) => config.name).join(', ');
            const subdomainConfig = { [subdomain]: getConfigObj(execPath, env, true) };
            const config = { mcpServers: subdomainConfig };
            const stringifiedConfig = JSON.stringify(subdomainConfig, undefined, 2);
            const configNumLines = stringifiedConfig.split('\n').length;
            if (numUpdates > 0) {
                log(`Successfully installed MCP server to ${namesOfSelectedConfigs} with the following configuration:\n${colors.dim}${JSON.stringify(config, undefined, 2)
                    .split('\n')
                    .map((line, index) => index > 1 && index < configNumLines
                    ? `${colors.default}${colors.green}+ ${line}${colors.default}${colors.dim}`
                    : `  ${line}`)
                    .join('\n')
                    .replace('\n    }', `\n${colors.default}${colors.dim}    }`)}`, 'success', LOGGING_CONFIG);
            }
        }
        catch (err) {
            log(`Error modifying config files with MCP servers:\n${err instanceof Error ? err.message : JSON.stringify(err, undefined, 2)}`, 'error', LOGGING_CONFIG);
        }
    });
}
export function getClaudeConfigPath() {
    const homeDir = os.homedir();
    const platform = process.platform;
    switch (platform) {
        case 'darwin':
            return path.join(homeDir, 'Library', 'Application Support', 'Claude');
        case 'win32':
            const appData = process.env.APPDATA;
            if (!appData) {
                throw new Error('APPDATA environment variable is not defined.');
            }
            return path.join(appData, 'Claude', 'claude_desktop_config.json');
        case 'linux':
            return path.join(homeDir, '.config', 'Claude');
        default:
            throw new Error(`Unsupported platform: ${platform}`);
    }
}
