// @mintlify/mcp/bin/cli.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { colors, log } from '@mintlify/scraping';
import { exec } from 'node:child_process';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { isMcpMetadata } from './remote.js';
import { formatMcpMetadata, getTerminalWidth, LOGGING_CONFIG, onCancel, padString, wrapText, } from './utils.js';
import { buildMcp } from './utils/cli-config.js';
import { modifyAllMcpConfigs } from './utils/mcp-configs.js';
process.on('SIGINT', onCancel);
process.on('SIGTERM', onCancel);
export const cli = () => yargs(hideBin(process.argv))
    .command('add <subdomain>', 'Add a new subdomain', (yargs) => {
    return yargs
        .positional('subdomain', {
        describe: 'The subdomain to install the MCP server from',
        type: 'string',
        demandOption: true,
    })
        .option('client', {
        describe: 'Specify which configs to modify',
        type: 'string',
        choices: ['cursor', 'windsurf', 'claude'],
        array: true,
    });
}, (argv) => __awaiter(void 0, void 0, void 0, function* () {
    const { subdomain, client } = argv;
    const configsToModify = client && client.length ? client : undefined;
    try {
        const homeDir = os.homedir();
        const mcpBaseDir = path.join(homeDir, '.mcp');
        const mcpDir = path.join(mcpBaseDir, subdomain);
        if (!fs.existsSync(mcpBaseDir)) {
            fs.mkdirSync(mcpBaseDir, { recursive: true });
        }
        const mcpMetadata = { source: 'mintlify' };
        if (isMcpMetadata(mcpMetadata)) {
            log(`Successfully fetched MCP for ${colors.green}${subdomain}${colors.default} subdomain:
${formatMcpMetadata(mcpMetadata)}`, 'info', { omitStatusMessage: true });
        }
        const filename = fileURLToPath(import.meta.url);
        yield buildMcp(mcpDir, subdomain, filename);
        log(`Created new MCP server at ${colors.green}${mcpDir}${colors.default}`, 'success', LOGGING_CONFIG);
        const gitignorePath = path.join(mcpDir, '.gitignore');
        fs.writeFileSync(gitignorePath, 'node_modules/');
        const execPath = path.join(mcpDir, 'src', `index.js`);
        try {
            process.chdir(mcpDir);
            exec('npm install');
            exec('git init');
            yield modifyAllMcpConfigs(subdomain, execPath, configsToModify);
        }
        catch (err) {
            log(`Error running ${colors.red}npm install && git init${colors.default}, please navigate to ${colors.blue}${mcpDir}${colors.default} and run ${colors.blue}npm install${colors.default} and initialize the git repo (if you'd like to save your changes)`, 'error', LOGGING_CONFIG);
        }
        finally {
            console.log();
            console.log(`${colors.default}To start your MCP server, run the following command:`);
            const rootCmd = 'node';
            const command = `${rootCmd} ${execPath}`;
            const cmdLen = command.length + 6;
            const terminalWidth = Math.min(getTerminalWidth(), cmdLen);
            const boxWidth = terminalWidth - 4;
            const contentWidth = boxWidth - 2;
            const topBorder = '┌' + '─'.repeat(boxWidth) + '┐';
            const bottomBorder = '└' + '─'.repeat(boxWidth) + '┘';
            let result = topBorder + '\n';
            const styledCmd = `${colors.green}${colors.bold}${command}${colors.default}`;
            const lines = wrapText(styledCmd, contentWidth);
            for (const line of lines) {
                result += `│ ${padString(line !== null && line !== void 0 ? line : '', contentWidth)} │\n`;
            }
            result += bottomBorder;
            console.log(result);
        }
    }
    catch (err) {
        const msg = err instanceof Error ? err.message : err;
        log(`${typeof msg === 'string' && (msg.startsWith('Failed to') || msg.startsWith('Error '))
            ? ''
            : `Error installing MCP for ${colors.red}${subdomain}${colors.default}: `}${err instanceof Error ? err.message : err}`, 'error', LOGGING_CONFIG);
    }
}))
    .help().argv;
