// @mintlify/mcp/bin/utils.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { colors, log } from '@mintlify/scraping';
export const LOGGING_CONFIG = {
    leadingNewLine: true,
    omitStatusMessage: true,
};
export function sleep(ms) {
    return __awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve) => setTimeout(resolve, ms));
    });
}
export function onCancel() {
    log('Exiting MCP installation', 'failure', LOGGING_CONFIG);
    process.exit(1);
}
export function generatePackageJson(subdomain) {
    return {
        name: `${subdomain}-mcp`,
        version: '1.0.0',
        type: 'module',
        description: `MCP server generated for the ${subdomain} subdomain from Mintlify`,
        engines: {
            node: '>=18.0.0',
        },
        main: `src/index.js`,
        scripts: {
            start: 'node src/index.js',
        },
        author: 'Mintlify',
        dependencies: {
            '@mintlify/openapi-types': '^0.0.0',
            '@mintlify/validation': '^0.1.320',
            '@modelcontextprotocol/sdk': '^1.6.1',
            axios: '^1.8.3',
            dashify: '^2.0.0',
            'js-yaml': '^4.1.0',
            'trieve-ts-sdk': '^0.0.62',
        },
    };
}
export function getTerminalWidth() {
    return Math.min(process.stdout.columns || 80, 80);
}
export function getVisibleString(str) {
    return str.replace(/\u001b\[.*?m/g, '');
}
function visibleLength(str) {
    return getVisibleString(str).length;
}
export function padString(str, width) {
    const visLen = visibleLength(str);
    return str + ' '.repeat(Math.max(0, width - visLen));
}
export function wrapText(text, maxWidth) {
    const lines = [];
    const paragraphs = text.split('\n');
    for (const paragraph of paragraphs) {
        if (paragraph.trim() === '') {
            lines.push('');
            continue;
        }
        const indentMatch = paragraph.match(/^(\s*)/);
        const indent = indentMatch ? indentMatch[1] : '';
        const indentLength = visibleLength(indent !== null && indent !== void 0 ? indent : '');
        const adjustedMaxWidth = maxWidth - indentLength;
        const textContent = paragraph.substring(indentLength);
        if (textContent.trim() === '') {
            lines.push(indent);
            continue;
        }
        let currentLine = indent;
        for (const word of textContent.split(' ')) {
            const nextPart = currentLine === indent ? word : ' ' + word;
            if (visibleLength(currentLine + nextPart) <= maxWidth) {
                currentLine += nextPart;
            }
            else {
                if (currentLine !== indent) {
                    lines.push(currentLine);
                    currentLine = indent;
                }
                if (visibleLength(indent + word) > maxWidth) {
                    let remainingWord = word;
                    while (remainingWord.length > 0) {
                        const chunk = remainingWord.substring(0, adjustedMaxWidth);
                        lines.push(indent + chunk);
                        remainingWord = remainingWord.substring(adjustedMaxWidth);
                    }
                    currentLine = indent;
                }
                else {
                    currentLine = indent + word;
                }
            }
        }
        if (currentLine !== indent) {
            lines.push(currentLine);
        }
    }
    return lines.filter((str) => str !== undefined);
}
export function formatMcpMetadata(metadata) {
    const terminalWidth = getTerminalWidth();
    const boxWidth = terminalWidth - 4;
    const contentWidth = boxWidth - 2;
    const nameLine = `${colors.bold}${colors.green}${metadata.name}${colors.default}`;
    const descriptionLines = wrapText(metadata.description, contentWidth);
    const date = `${colors.italic}${colors.dim}Last Updated: ${new Date(metadata.updatedAt).toDateString()}${colors.default}`;
    const authorLine = `${colors.blue}Author:${colors.default} ${metadata.author.name}`;
    const categoriesLine = `${colors.blue}Categories:${colors.default} ${metadata.categories.join(', ')}`;
    const verifiedBadge = metadata.verified === 'VERIFIED' ? `${colors.green}✓${colors.default}` : ' ';
    const topBorder = '┌' + '─'.repeat(boxWidth) + '┐';
    const bottomBorder = '└' + '─'.repeat(boxWidth) + '┘';
    let result = topBorder + '\n';
    result += `│ ${padString(nameLine, boxWidth - 4) + verifiedBadge}  │\n`;
    for (const line of descriptionLines) {
        result += `│ ${padString(line !== null && line !== void 0 ? line : '', contentWidth)} │\n`;
    }
    result += `│ ${padString(date, contentWidth)} │\n`;
    result += `│ ${' '.repeat(contentWidth)} │\n`;
    const authorLines = wrapText(authorLine, contentWidth);
    for (const line of authorLines) {
        result += `│ ${padString(line !== null && line !== void 0 ? line : '', contentWidth)} │\n`;
    }
    const categoriesLines = wrapText(categoriesLine, contentWidth);
    for (const line of categoriesLines) {
        result += `│ ${padString(line !== null && line !== void 0 ? line : '', contentWidth)} │\n`;
    }
    result += bottomBorder;
    return result;
}
