// @mintlify/mcp/src/mcp-server/src/tools/helpers.ts
import { OpenAPI } from '@mintlify/openapi-types';
import {
  HttpMethod,
  OpenApiToEndpointConverter,
  SecurityParameterGroup,
  SecurityParameterSections,
} from '@mintlify/validation';
import { Endpoint } from '@mintlify/validation';
import { ParameterSections } from '@mintlify/validation';
import { IncrementalDataSchemaArray } from '@mintlify/validation';
import { ParameterGroup } from '@mintlify/validation';
import { DataSchemaArray } from '@mintlify/validation';
import { Tool } from '@modelcontextprotocol/sdk/types.js';
import dashify from 'dashify';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';
import { z } from 'zod';

import { ServerParams, ToolWithEndpoint } from '../types.js';
import { initializeObject, NestedRecord, SimpleRecord } from '../utils.js';
import { dataSchemaArrayToZod, dataSchemaToZod } from './zod.js';

export function convertStrToTitle(str: string): string {
  const spacedString = str.replace(/[-_]/g, ' ');
  const words = spacedString.split(/(?=[A-Z])|\s+/);
  const titleCasedWords = words.map((word) => {
    return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
  });
  return titleCasedWords.join(' ');
}

export function findNextIteration(set: Set<string>, str: string): number {
  let count = 1;
  set.forEach((val) => {
    if (val.startsWith(`${str}---`)) {
      count = Number(val.replace(`${str}---`, ''));
    }
  });
  return count + 1;
}

export function getMcpEnabledEndpointsFromOpenApiSpec(
  spec: OpenAPI.Document
): Endpoint<DataSchemaArray>[] {
  const mcpEnabledEndpoints = [];

  const isMcpEnabledGloballyInSpec = spec['x-mcp']?.enabled === true;
  const endpoints = getEndpointsFromOpenApi(spec);

  if (isMcpEnabledGloballyInSpec) {
    const notDisabledEndpoints = endpoints.filter((endpoint) => endpoint.xMcp?.enabled !== false);
    mcpEnabledEndpoints.push(...notDisabledEndpoints);
  } else {
    const enabledEndpoints = endpoints.filter((endpoint) => endpoint.xMcp?.enabled === true);
    mcpEnabledEndpoints.push(...enabledEndpoints);
  }

  return mcpEnabledEndpoints;
}

export function convertEndpointToTool(
  endpoint: Endpoint<DataSchemaArray>
): Omit<Tool, 'inputSchema'> {
  let name;

  if (endpoint.xMcp?.name) {
    name = endpoint.xMcp.name;
  } else if (endpoint.title) {
    name = dashify(endpoint.title);
  } else {
    name = convertStrToTitle(endpoint.path);
  }

  let description;

  if (endpoint.xMcp?.description) {
    description = endpoint.xMcp.description;
  } else if (endpoint.description) {
    description = endpoint.description;
  } else {
    description = `${endpoint.method} ${endpoint.path}`;
  }

  return {
    name,
    description,
  };
}

export function getMcpToolsAndEndpointsFromOpenApiSpec(spec: OpenAPI.Document): ToolWithEndpoint[] {
  const endpoints = getMcpEnabledEndpointsFromOpenApiSpec(spec);

  const toolsWithEndpoints: ToolWithEndpoint[] = [];

  endpoints.forEach((endpoint) => {
    const tool = convertEndpointToTool(endpoint);
    toolsWithEndpoints.push({
      tool,
      endpoint,
    });
  });

  return toolsWithEndpoints;
}

export function getEndpointsFromOpenApi(specification: OpenAPI.Document) {
  const endpoints = [];
  const paths = specification.paths;

  for (const path in paths) {
    const pathObj = paths[path];
    const httpMethods = Object.values(OpenAPIV3.HttpMethods);

    for (const method of httpMethods) {
      if (!pathObj || !(method in pathObj)) {
        continue;
      }
      const endpoint = OpenApiToEndpointConverter.convert(
        specification as OpenAPIV3_1.Document,
        path,
        method as HttpMethod,
        true
      );
      endpoints.push(endpoint);
    }
  }

  return endpoints;
}

export function loadEnv(key: string): SimpleRecord {
  let envVars: SimpleRecord = {};
  try {
    const envPath = path.join(fileURLToPath(import.meta.url), '../../../', '.env.json');
    if (fs.existsSync(envPath)) {
      envVars = JSON.parse(fs.readFileSync(envPath).toString());
      return (envVars[key] as SimpleRecord | undefined) ?? {};
    }
  } catch (error) {
    // if there's no env, the user will be prompted
    // for their auth info at runtime if necessary
    // (shouldn't happen either way)
  }
  return envVars;
}

function convertParameterSection(
  parameters: ParameterGroup<DataSchemaArray | IncrementalDataSchemaArray>,
  paramSection: ServerParams
): void {
  Object.entries(parameters).forEach(([key, value]) => {
    const schema = value.schema;
    paramSection[key] = dataSchemaArrayToZod(schema);
  });
}

function convertParametersAndAddToRelevantParamGroups(
  parameters: ParameterSections,
  paths: ServerParams,
  queries: ServerParams,
  headers: ServerParams,
  cookies: ServerParams
): void {
  convertParameterSection(parameters.path, paths);
  convertParameterSection(parameters.query, queries);
  convertParameterSection(parameters.header, headers);
  convertParameterSection(parameters.cookie, cookies);
}

// this function returns all the securityParameters and seeds them with the envVariables if we have them
export function convertSecurityParameterSection(
  securityParameters: SecurityParameterGroup,
  envVariables: SimpleRecord,
  location: string
): {
  key: string;
  value: NestedRecord | undefined;
}[] {
  const res: {
    key: string;
    value: NestedRecord | undefined;
  }[] = [];
  Object.entries(securityParameters).forEach(([key, value]) => {
    let envKeyList: string[] = [];
    let targetKey = '';
    switch (value.type) {
      case 'apiKey':
        envKeyList = [location, key];
        targetKey = 'API_KEY';
        break;
      case 'http':
        envKeyList = [location, key, 'HTTP'];
        targetKey = value.scheme;
        break;
      case 'oauth2':
      default:
        break;
    }

    const target = initializeObject(
      { ...(envVariables as Record<string, SimpleRecord>) },
      envKeyList
    );
    if (envKeyList.length && !target[targetKey]) {
      res.push({
        key,
        value: undefined,
      });
    } else {
      res.push({
        key,
        value: target[targetKey],
      });
    }
  });

  return res;
}

function convertSecurityParametersAndAddToRelevantParamGroups(
  securityParameters: SecurityParameterSections,
  queries: ServerParams,
  headers: ServerParams,
  cookies: ServerParams,
  envVariables: SimpleRecord
): void {
  const queryRes = convertSecurityParameterSection(securityParameters.query, envVariables, 'query');
  const headerRes = convertSecurityParameterSection(
    securityParameters.header,
    envVariables,
    'header'
  );
  const cookieRes = convertSecurityParameterSection(
    securityParameters.cookie,
    envVariables,
    'cookie'
  );

  // non intuitive that we seed the query with a zod type if it *doesn't* exist
  // but that's because if we don't have it in our env Variables, that means the user should provide it in their query.
  queryRes.forEach(({ key, value }) => {
    if (!value) {
      queries[key] = z.string();
    }
  });
  headerRes.forEach(({ key, value }) => {
    if (!value) {
      headers[key] = z.string();
    }
  });
  cookieRes.forEach(({ key, value }) => {
    if (!value) {
      cookies[key] = z.string();
    }
  });
}

export function convertEndpointToCategorizedZod(
  envKey: string,
  endpoint: Endpoint
): {
  url: string;
  method: HttpMethod;
  paths: ServerParams;
  queries: ServerParams;
  body: ServerParams | undefined;
  headers: ServerParams;
  cookies: ServerParams;
} {
  const envVariables = loadEnv(envKey);
  const url = `${endpoint.servers?.[0]?.url || ''}${endpoint.path}`;
  const method = endpoint.method;
  const paths: ServerParams = {};
  const queries: ServerParams = {};
  const headers: ServerParams = {};
  const cookies: ServerParams = {};
  let body: ServerParams | undefined = undefined;

  convertParametersAndAddToRelevantParamGroups(
    endpoint.request.parameters,
    paths,
    queries,
    headers,
    cookies
  );

  if (endpoint.request.security[0]?.parameters) {
    convertSecurityParametersAndAddToRelevantParamGroups(
      endpoint.request.security[0].parameters,
      queries,
      headers,
      cookies,
      envVariables
    );
  }

  const jsonBodySchema = endpoint.request.body['application/json'];
  const bodySchemaArray = jsonBodySchema?.schemaArray;
  const bodySchema = bodySchemaArray?.[0];

  if (bodySchema) {
    const zodBodySchema = dataSchemaToZod(bodySchema);
    body = { body: zodBodySchema };
  }

  return { url, method, paths, queries, body, headers, cookies };
}
