// @mintlify/mcp/src/mcp-server/src/tools/index.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import axios, { isAxiosError } from 'axios';
import dashify from 'dashify';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { ToolWithEndpointWithUuid } from '../types.js';
import {
  convertEndpointToCategorizedZod,
  convertSecurityParameterSection,
  convertStrToTitle,
  findNextIteration,
  loadEnv,
} from './helpers.js';

export async function createTools(server: McpServer, existingTools: Set<string>) {
  const toolsDir = path.join(fileURLToPath(import.meta.url), '..', '..');
  let tools = JSON.parse(fs.readFileSync(path.join(toolsDir, 'tools.json'), 'utf-8'));

  tools = tools.filter((tool: ToolWithEndpointWithUuid) => tool.endpoint);

  tools.forEach(({ uuid, endpoint }: ToolWithEndpointWithUuid) => {
    const envVars = loadEnv(uuid);

    const {
      url: urlSchema,
      method: methodSchema,
      paths: pathsSchema,
      queries: queriesSchema,
      body: bodySchema,
      headers: headersSchema,
      cookies: cookiesSchema,
    } = convertEndpointToCategorizedZod(uuid, endpoint);

    const serverArgumentsSchemas = {
      ...pathsSchema,
      ...queriesSchema,
      ...bodySchema,
      ...headersSchema,
      ...cookiesSchema,
    };

    if (!endpoint.title) {
      endpoint.title = `${endpoint.method} ${convertStrToTitle(endpoint.path)}`;
    }

    if (existingTools.has(endpoint.title)) {
      const lastCount = findNextIteration(existingTools, endpoint.title);
      endpoint.title = `${endpoint.title}---${lastCount}`;
    }

    if (endpoint.title.length > 64) {
      endpoint.title = endpoint.title.slice(0, -64);
    }

    existingTools.add(endpoint.title);

    server.tool(
      dashify(endpoint.title),
      endpoint.description || endpoint.title,
      serverArgumentsSchemas,
      async (inputArgs) => {
        const inputParams: Record<string, string> = {};
        const inputHeaders: Record<string, string> = {};
        const inputCookies: Record<string, string> = {};

        let urlWithPathParams = urlSchema;

        let inputBody: unknown | undefined = undefined;
        if ('body' in inputArgs) {
          inputBody = inputArgs.body;
          delete inputArgs.body;
        }

        Object.entries(inputArgs).forEach(([key, value]) => {
          if (key in pathsSchema) {
            urlWithPathParams = urlWithPathParams.replace(`{${key}}`, value);
          } else if (key in queriesSchema) {
            inputParams[key] = value;
          } else if (key in headersSchema) {
            inputHeaders[key] = value;
          } else if (key in cookiesSchema) {
            inputCookies[key] = value;
          }
        });

        const securityParamSections = endpoint.request.security[0]?.parameters;

        if (securityParamSections) {
          const headerRes = convertSecurityParameterSection(
            securityParamSections.header,
            envVars,
            'header'
          );
          headerRes.forEach(({ key, value }) => {
            if (value) {
              inputHeaders[key] = value as string;
            }
          });
          const cookieRes = convertSecurityParameterSection(
            securityParamSections.cookie,
            envVars,
            'cookie'
          );
          cookieRes.forEach(({ key, value }) => {
            if (value) {
              inputCookies[key] = value as string;
            }
          });
          const queryRes = convertSecurityParameterSection(
            securityParamSections.query,
            envVars,
            'query'
          );

          queryRes.forEach(({ key, value }) => {
            if (value) {
              inputParams[key] = value as string;
            }
          });
        }

        try {
          const response = await axios({
            url: urlWithPathParams,
            method: methodSchema,
            params: inputParams,
            data: inputBody,
            headers: inputHeaders,
          });

          return {
            content: [
              {
                type: 'text',
                text: JSON.stringify(response.data, undefined, 2),
              },
            ],
          };
        } catch (error) {
          const errMsg = JSON.stringify(error, undefined, 2);
          return {
            isError: true,
            content: [
              {
                type: 'text',
                text: isAxiosError(error) ? `${error.message}\n\n${errMsg}` : errMsg,
              },
            ],
          };
        }
      }
    );
  });
}
