// @mintlify/mcp/src/remote.ts
import axios from 'axios';
import { Readable } from 'node:stream';
import { z } from 'zod';

import { extractTarGzStream } from './compress.js';
import { throwOnAxiosError } from './mcp-server/src/utils.js';

/**
 * @deprecated MCPT backend to be sunset
 */
const MCPT_SERVER_URL = `https://server.mcpt.com/api/v1/public`;
const MCPT_METADATA_ROUTE = '/server/';

export const McpAuthorSchema = z.object({
  id: z.number(),
  createdAt: z.string(),
  updatedAt: z.string().optional(),
  name: z.string(),
  logo: z.string(),
  href: z.string().url().nullable(),
});

// TODO (Ricardo): Add better typing here for tools
export const ToolSchema = z.record(z.string(), z.unknown());

export const McpMetadataSchema = z.object({
  id: z.string(),
  createdAt: z.string(),
  updatedAt: z.string(),
  source: z.string(),
  slug: z.string(),
  name: z.string(),
  logo: z.string(),
  authorId: z.number(),
  author: McpAuthorSchema,
  description: z.string(),
  instructions: z.string(),
  verified: z.string(),
  categories: z.array(z.string()),
  documentationHref: z.string().nullable(),
  tools: z.array(ToolSchema),
  presignedUrl: z.string().optional(),
  mcpConfiguration: z.record(
    z.string(),
    z.object({
      env: z.record(z.string(), z.string()),
    })
  ),
});

export type McpAuthor = z.infer<typeof McpAuthorSchema>;
export type Tool = z.infer<typeof ToolSchema>;
export type McpMetadata = z.infer<typeof McpMetadataSchema>;

export function isMcpMetadata(metadata: unknown): metadata is McpMetadata {
  return McpMetadataSchema.safeParse(metadata).success;
}

/**
 * @deprecated MCPT backend to be sunset
 * @param subdomain
 * @returns MCP metadata. and a presigned link to the MCP bundle
 */
export async function fetchMcp(subdomain: string): Promise<McpMetadata> {
  const response = await axios.get(`${MCPT_SERVER_URL}${MCPT_METADATA_ROUTE}${subdomain}`, {
    headers: {
      'X-MCPT-Source': 'cli',
    },
    validateStatus() {
      return true;
    },
  });
  throwOnAxiosError(response, 'Failed to fetch MCP metadata');
  return response.data;
}

/**
 * @deprecated MCPT backend to be sunset
 * @param presignedUrl
 * @param destinationPath
 * @param subdomain
 */
export async function unzipMcp(
  presignedUrl: string,
  destinationPath: string,
  subdomain: string
): Promise<void> {
  const response = await axios({
    method: 'get',
    url: presignedUrl,
    headers: {
      Accept: 'application/x-gzip',
    },
    responseType: 'arraybuffer',
    validateStatus() {
      return true;
    },
  });
  throwOnAxiosError(response, 'Failed to fetch MCP bundle');
  const buf = Buffer.from(response.data);
  const stream = new Readable();
  stream.push(buf);
  stream.push(null);
  await extractTarGzStream(stream, destinationPath, subdomain);
}
